### PR TITLE
Adds asset-ref relation to Asset graphql schema

### PR DESCRIPTION
**Pull request by agzam**
Just like every entity can relate to multiple [Feedback records](https://github.com/threatgrid/ctia/blob/master/src/ctia/entity/feedback/graphql_schemas.clj#L33), every Asset record can be associated with zero, one, or multiple AssetMappings and/or AssetProperties.

> **Epic** #
> Close https://github.com/threatgrid/iroh/issues/4521
> Related https://github.com/threatgrid/ctia/pull/1064

*This the second part of #1064* 

<a name="qa">[§](#qa)</a> QA
============================

1. Create a few `Assets` 
2. Create a few `AssetMappings` with `asset_ref` pointing to `Assets`, preferably multiple `AssetMappings` per single `Asset`. Note that you'd have to use `Asset` and `AssetMapping` POST requests separately. At the moment Bundle Importing Assets and related AssetMappings using transient IDs in asset_ref is [not implemented yet](https://github.com/threatgrid/iroh/issues/4751). That means you'd have to create an Asset (`POST /ctia/asset`), write down its ID; Create AssetMapping(s) (`POST /ctia/asset-mapping`) with `asset_ref` to be of that ID, and so on.
3. Similarly, create multiple `AssetProperties` per each `Asset` object
4. Test by sending GraphQL queries  to `/ctia/graphql` (for each created Asset) like this:
```
{
    "query": "query Asset { asset(id:\"http://localhost:3000/ctia/asset/asset-61c378d5-c38d-45e1-a6f0-590d44e9cac8\") {id asset_mappings{ totalCount nodes { asset_ref }}}}"
}
    
```
    
  That query should return a map such this:
   
    {
      "data": {
        "asset": {
          "id": "http://localhost:3000/ctia/asset/asset-61c378d5-c38d-45e1-a6f0-590d44e9cac8",
          "asset_mappings": {
            "totalCount": 1,
            "nodes": [
              {
                "asset_ref": "http://localhost:3000/ctia/asset/asset-61c378d5-c38d-45e1-a6f0-590d44e9cac8"
              }
            ]
          }
        }
      }
    }

5. Make sure that `totalCount` is matching the number of `AssetMappings` created per `Asset`
6. Make sure that all `asset_ref` fields have the same `AssetId`
7. You can test `AssetProperties` similarly:

```
{
    "query": "query Asset { asset(id:\"http://localhost:3000/ctia/asset/asset-61c378d5-c38d-45e1-a6f0-590d44e9cac8\") {id asset_properties{ totalCount nodes { asset_ref }}}}"
}
```

<a name="squashed-commits">[§](#squashed-commits)</a> Squashed Commits
======================================================================

5b7c4ee6 change rand-btw to take half-open interval
21934579 test: asset graphql asset_ref test
dc3aad56 fix resolvers for asset-mapping and asset-properties
253a2ed5 conjoin graphql queries for Asset, AssetMapping and AssetProperties
6b1901f6 fix for: `Verdict not found in schema`
3b645da8 adds asset-ref relation to asset graphql schema
b12027f8 remove singular form of asset-properties
5ef88850 adds AssetProperties tests
e37c635d Adds graphql schemas for asset-mapping

### CODE REVIEW

**Comment by agzam** [comment]
@ereteog I'm stuck again with the same error I reported in https://github.com/threatgrid/ctia/pull/1064#discussion_r568236199

```
graphql.AssertException: type Verdict not found in schema
```

Only this time it seems I'm stuck for real. I don't know what it wants. The moment I add this line: https://github.com/threatgrid/ctia/blob/asset_graphql_schema__asset_ref__relationship_%234521/src/ctia/entity/asset/graphql_schemas.clj#L74

It doesn't work anymore. Also, for some reason, I have to restart the REPL (removing the line, re-evaluating doesn't set the state right)

**Comment by ereteog** [comment]
Some remark about your statements in QA section (I corrected a part):

> Bulk Importing Assets and related AssetMappings using transient IDs not working properly

Bulk import is not supposed to resolve any transient ids. It is just an helper to create multiple instances at once. Thus I assume that your are referring to bundle import which is effectively an helper to import many entities and link them through relationships source target. The entities are not created yet so we do not know yet their ids to use in relationships, that's why we use transient ids. So, under the hood, [bundle import](https://github.com/threatgrid/ctia/blob/f5effb8e687973af89245c219952ba0773a83b6a/src/ctia/bundle/core.clj#L295) generates first the entities (evenutally retrieve existing one if any of the `external_ids` values are matched), map the ids of these new entities to transient ids, resolve transient ids in relationships before creating it, and then submit it to the bulk fn.
I created a [task](https://github.com/threatgrid/iroh/issues/4751) to implement the same mechanism but by resolving  transient ids in `asset_ref` too.

> That means you'd have to create an Asset (POST /ctia/asset), write down its ID; Create AssetMapping(s) (POST /ctia/asset-mapping) with asset_ref to be of that ID, and so on.

A more optimal approach is to create multiple Assets with a bulk import, retrieve the ids to use in `asset-mapping`, `asset-properties` and `target-record` entities and submit them in a second bulk.


To summarize:
- bulk is as an helper to import/export many entities at once. 
- bundle is a bulk helper that resolves external ids to avoid duplicates and transient ids to submit entities and their relationship at once (ex: [describe an incident in a bundle](https://github.com/threatgrid/ctim/blob/master/doc/tutorials/modeling-incidents-in-ctim.md))
- we will implement a bundle helper to resolve transient ids in `asset_ref` fields

**Review by ereteog — APPROVED** [review]
Nice!

File: test/ctia/http/routes/graphql/asset_test.clj [hunk 1]
```diff
@@ -85,3 +87,64 @@
...
+  (->> assets
+       (mapcat
+        (fn [{:keys [id]}]
+          (create-random-objects
+           app
+           (rand-btw 1 3)
+           (-> example
+               (dissoc :id)
+               (assoc :asset_ref id)))))
+       (into [])))
```
[the remark this answers is not in the document]
Comment (agzam) [reply test/ctia/http/routes/graphql/asset_test.clj:118]: because otherwise it's evaluated lazily and won't create asset-maps or asset-properties.

File: test/ctia/http/routes/graphql/asset_test.clj [hunk 2]
```diff
@@ -85,3 +87,64 @@
              (is (= [asset2]
                     (get-in data [:assets :nodes]))
                  "The asset matches the search query"))))))))
+
+(defn- rand-btw
+  "Generates random int between two numbers"
+  [a n] (+ a (rand-int (- (inc n) a))))
```
[the remark this answers is not in the document]
Comment (agzam) [reply test/ctia/http/routes/graphql/asset_test.clj:93]: > ranges generally includes the lower bound but excludes the upper bound

Ah, my bad. For some reason I was thinking it should be a "closed interval" (as in Math), I was trying with small numbers and forgot that `clojure.core/range` does exactly as you described. Will fix it.

File: test/ctia/http/routes/graphql/asset_test.clj [hunk 3]
```diff
@@ -85,3 +87,64 @@
...
+           (-> example
+               (dissoc :id)
+               (assoc :asset_ref id)))))
+       (into [])))
+
+(deftest asset-graphql-asset-refs-test
+  (test-for-each-store-with-app
+   (fn [app]
+     (helpers/set-capabilities! app "foouser" ["foogroup"] "user" all-capabilities)
+     (whoami-helpers/set-whoami-response app "45c1f5e3f05d0" "foouser" "foogroup" "user")
```
[the remark this answers is not in the document]
Comment (agzam) [reply test/ctia/http/routes/graphql/asset_test.clj:124]: It doesn't work for me locally when I remove any of these lines.

**Comment by hisans** [comment]
QA Verified in 1.66 Release
